### ISSUE TITLE
Enable Go to Implementation in Typescript Editors

### ISSUE THREAD

**Issue by doliver3** [microsoft/TypeScript#6209]
Could you provide a way for Typescript Editors to list the implementations of a type?
This would definitely help developer productivity to be able to find implementations of types faster. 

For example, in the Atom Editor atom-typescript plugin you can press F12 to 'Go to Declaration' which as currently implemented you often arrive at the typescript .d.ts file which has the function signature but not the actual implementation. 

Since the Typescript parser often generates the d.ts files from the .ts implementation files. Could that association between the d.ts and .ts files be preserved (perhaps output with a typescript compiler option) so that Typescript editors could take developers to a list of the implementing classes or functions?

I opened a similar request for the Atom-typescript editor, but they indicate that they cannot implement this without Typescript parser support:
TypeStrong/atom-typescript#790

**Comment by DanielRosenwasser** [comment]
So you want this functionality to jump to the corresponding `.js` if available, provided that the definition is ambient, correct?

**Comment by DanielRosenwasser** [comment]
@RyanCavanaugh @mousetraps @bowdenk7

**Comment by doliver3** [comment]
> So you want this functionality to jump to the corresponding .js if available, provided that the definition is ambient, correct?
Yes this is a scenario that would be very useful for productivity. This alone would be very valuable. I believe the atom-typescript author believes more hooks in the Typescript compiler would make it easier for him to implement this.  I'm unclear on whether this is information on js implementation line numbers is information the typescript compiler would typically know internally.

In addition, another scenario would be enabling a jump to the original typescript .ts file definition in the library being used by the application. I would imagine this information is available pre-compile to d.ts and .js files.  We now have entire sets of libraries being written in .ts files as source outputting ES5 syntax js. It would be nice to achieve the ability to jump to the original typescript implementation definition in the .ts files. 

I'm aware this is asking even more to jump to the .ts file implementation in the required library as it might need source map usage or distribution of source/link to original ts source.   It would seem to me that the typescript compiler that generates the initial library of functions and classes would know the lines where the implementation is at and could preserve the line numbers when the library is distributed somehow so that the typscript compiler of the application using the class could know what line to jump to in the library.  

Is this information on line number of implementing function/class available in source maps?  Are there other sources of this information?

**Comment by DanielRosenwasser** [comment]
Given that we are going to be using the `--allowJS` flag, the language service should be able to work with  the JS files.

I don't know how `.d.ts` files and `.js` files currently work together in that respect. @sheetalkamat probably (definitely) knows more about this than I do.

**Comment by riknoll** [comment]
Reopening, because going to js files isn't supported quite yet

**Comment by ubershmekel** [comment]
Any updates on this one? The current state of typescript is that you cannot jump to the implementation of code in many prominent libraries.

This issue has been moved/closed/duped/deduped multiple times and fixing this bug could be a huge productivity boost.

https://github.com/ubershmekel/vscode-ts-goto-source has a project that reproduces the issue. As explained in https://github.com/Microsoft/vscode/issues/26325 and referenced in https://github.com/Microsoft/vscode/issues/10806

Note I had to use `"typescript.disableAutomaticTypeAcquisition": true` from one point to use the javascript "go to definition" because it seems the `.d.ts` files take over if found.

**Comment by mmc41** [comment]
@DanielRosenwasser Any updates on this? - Single most wanted feature for me.

**Comment by Pajn** [comment]
This is particularily annoying if working in a monorepo written entirely in Typescript. I have multiple times when adding a field to a type or similar smaller changes used "go to definition", added it and then later found it missing again realizing that I added it to the `.d.ts` file and not the actual `.ts` file

**Comment by ziozxzioz** [comment]
@DanielRosenwasser Is this progressing? I would really love this feature

**Comment by mrdulin** [comment]
same issue. Want to "Go to implementation" to the source code, not `xx.d.ts`

**Comment by Pajn** [comment]
>going to source is always more preferable IMHO (if it is available)

When the source is in TS. If the source is in JS though there's a good reason to go to the .d.ts

**Comment by dakaraphi** [comment]
>When the source is in TS. If the source is in JS though there's a good reason to go to the .d.ts

Why? I can just hover to see the type definition.  Anyway, there should be a separate command for type definition vs implementation.

**Comment by ivmos** [comment]
Is there any reason why this is still pending? A workaround, alternative or something? Thanks in advance

**Comment by weswigham** [comment]
> Since the Typescript parser often generates the d.ts files from the .ts implementation files. Could that association between the d.ts and .ts files be preserved (perhaps output with a typescript compiler option) so that Typescript editors could take developers to a list of the implementing classes or functions?

I have a [branch](https://github.com/weswigham/TypeScript/tree/declaration-maps) of #21930 that enables producing precisely these declaration file sourcemaps (and will have that out for review either on request or once #21930 is merged), and am working on adding support into our LS for following these maps. We want to ship it in 2.8.

So yeah, we've been working on it. It just needed a lot of internal changes. 😄

**Comment by zheeeng** [comment]
Expect updates on this issue to relieve the pain on finding implementations through manually expand folders under node_modules and locate the entry module file.

**Comment by derN3rd** [comment]
Is there any workaround for this?
As we have a big module for our db queries written in ts it would be really nice to see the real source code with one command instead of searching through the node_modules.

**Comment by ewnd9** [comment]
@derN3rd There is [`vscode-search-node-modules`](https://github.com/jasonnutter/vscode-search-node-modules) extension to browse `node_modules` in the command panel (and I also have a fork which always opens README [`vscode-open-node-modules`](https://github.com/ewnd9/vscode-open-node-modules)) 

I understand it's far from the perfect ux to click on require/import and open source a file, but definitely better than scrolling `node_modules`

**Comment by weswigham** [comment]
@derN3rd we will be shipping a [`--declaratrionMap`](https://github.com/Microsoft/TypeScript/pull/22658) flag in 2.9 (which may be released during next month), and we've already enabled going to definitions in the original source TS over an associated `.d.ts` using them (which, if you're shipping declaration files in your node modules folder may get a lot of what you want). We're also looking into jumping to/from the associated JS as well, provided normal sourcemaps are also on.

**Comment by derN3rd** [comment]
@weswigham thanks for the fast reply! Sounds good
Could you post an update here if the feature is ready or ready to be tested?

**Comment by weswigham** [comment]
`--declarationMap` (and associated LS redirection) has been in our nightlies since just after the 2.8 release. The only bit we've not done yet is the JS file hopping, as that may require some extra editor coordination.

**Comment by derN3rd** [comment]
Any news for that on 3.0-rc?

**Comment by SheikhG1900** [comment]
Same issue with `.css` and `.json` files.  **Go to Definition** taking me to `.d.ts` file.
```
import React from 'react'
import Route from 'react-router-dom/Route'
import Switch from 'react-router-dom/Switch'
import Loadable from 'react-loadable'
import  './App.css'
import './tailwind.css'
```  
**Go to Definition** on `./App.css` takes me to

```
declare module '*.css' {
    const content: any;
    export default content;
}

declare module '*.svg' {
    const content: any;
    export default content;
}

declare module '*.json' {
    const content: any;
    export default content;
}
```

**Comment by derN3rd** [comment]
@robsman _Go to Definition_ should take you to the compiled js code, while _Go to Type Definition_ should take you to the ts file.
That issue still exists for me in VSCode 1.27.0, as both functions take me to the .ts file

**Comment by robsman** [comment]
@derN3rd re-tested, both _Go to Definition_ and _Go to Type Definition_ as well as _Peek Definition_ take me to the real code in the _.ts_ file.

The API declaration and implementation sits in a separate module that the calling package refers to as a _git+ssh..._ dependency.

**Comment by Zx-EvM** [comment]
@robsman I think `Go to Definition` should refer me to source of this class on function in js file (if it's javascript) inside `node_modules` not `.d.ts`. Anyway I don't understand for what VSCode this feature in the current state

**Comment by prufeng** [comment]
Hope can prioritize this issue, it is a really a big pain, from so many comments and related issues.

**Comment by phil-lgr** [comment]
Maybe I'm missing something, but the way I see the default CMD + Click behaviour is:
1. if the .d.ts exist go there
2. if it doesn't exist go to the actual js file

right now I can't click to navigate to a .js file

![image](https://user-images.githubusercontent.com/9644867/47965360-36504c80-e014-11e8-8eff-2dc7c0d38aac.png)

**Comment by bartlettpsj** [comment]
Please add me to the long list of folk trying to use command-click to navigate through the code. I have no TS in my project and the npm package I am navigating to does not use TS, yet command-click on require('co-body') takes me to an internally generated TS definition.   I am long time IntelliJ user and still, because of this, I fall back to IntelliJ.

**Comment by chaosforfun** [comment]
it's a long request

**Comment by restjohn** [comment]
I like VS Code a lot and I just started a new Node project and decided to use TypeScript for the first time.  So far so mostly good, but this issue has been a pain point.  Working in open-source, the ability to click through to real source code in dependencies is vital to work-flow.  I am working in TypeScript source files and using JS NPM modules with associated `@types/xxx` modules that have no associated documentation.  IntelliSense with TypeScript is great, but doesn't always provide enough information.  At the very least, being able to click straight through to the module source from the `import` statement would be nice.  I believe on some libraries that works, and on others, e.g. Passport, I can't get past the adjunct TypeScript stubs in the `@types` module.

**Comment by remcohaszing** [comment]
This should be labeled _Bug_, not _Suggestion_.

The lack of code navigation really ruins the VSCode experience for JavaScript developers.

**Comment by promaty** [comment]
I cannot believe this is open since 2015, how do you guys conveniently navigate to implementation?

**Comment by phil-lgr** [comment]
sorry to spam the thread but @promaty I use the [search-node-modules](https://marketplace.visualstudio.com/items?itemName=jasonnutter.search-node-modules) extension 😬 for now

**Comment by joeysino** [comment]
WebStorm has the same behaviour: [StackOverflow](https://stackoverflow.com/questions/42422033/webstorm-typescript-how-to-navigate-to-ts-instead-of-d-ts), [BugTracker](https://youtrack.jetbrains.com/issue/WEB-23618)

VSCode has "Go to Definition" and "Go to Type Definition" actions

WebStorm has "Go To Declaration" and "Go To Type Declaration" actions

But in TypeScript it seems they all do the same thing, go to the type definition!

**Comment by henrahmagix** [comment]
"Go to Definition" and "Go to Type Definition", when used on `node_modules/` libraries, both take me to `.d.ts` files. I'm on the latest VS Code as of writing (1.34.0).

For example: npm package `@angular/core@7.2.15`, when I import `TestBed` from `@angular/core/testing` and try to see the definition of `TestBed.createComponent`, I get two different things:
- "Go to Definition" -> `node_modules/@angular/core/testing/src/test_bed_common.d.ts` line 117
- "Go to Type Definition" -> `node_modules/@angular/core/testing/src/component_fixture.d.ts` line 14

This basically means that the Angular source code is hidden from me in VS Code 😞 I'm not sure if that's a symptom of how the Angular library is written, or how TypeScript/VS Code finds source implementations.

All I know is I can only reliably use "Go to Definition" and "Go to Type Definition" on my own code, in which case I get taken directly to the source; I'm never shown a `.d.ts` file for my own code.

**Comment by silverqx** [comment]
We have to manually search source codes for now :/

**Comment by Happin3ss** [comment]
Aw, I just figure out that somehow Intellij IDEA have this function work really well, even I'm on the same project, using Typescript 3.0.1

**Comment by kaelzhang** [comment]
I thought the priority of this issue should be bumped up

**Comment by hushicai** [comment]
So That's why `flow` is still useful.

**Comment by sheerun** [comment]
Could someone from typescript team suggest where in code this can be potentially fixed?

**Comment by sheerun** [comment]
A funny thing is that if some package doesn't have types available "Go To Type Definition" actually works better for it. It means that **not** using typings is better if you like "Go To" features 😄 

Here's screencast that shows that I'm able to go to definition of `device` packages that doesn't have typings, but `compression` leads me to some random typings file:

https://giphy.com/gifs/j3VCp0LVKr5LsMUh11

@RyanCavanaugh I must agree with other people, this should be labeled as bug, not suggestion..

**Comment by Ashnxious** [comment]
That is the case for me too, in "node_modules", `Go To Definition` and `Go to Type Definition` both send my to the `*.d.ts` file. The library is `tensorflow`

**Comment by tafelito** [comment]
in my case, if I have something like this defined in my own code

```ts
const Foo: <T> = () => {
...
}
```

if the T comes from a npm module, when I try to go to definition, it will show me 2 definitions. My own code and the typing definition. If I go to type definition it goes straight to the definition file. 
In the first case, it should go straight to the definition. 
To work around this, I changed the editor --> goto location: multiple to `goto` instead of `peek`

**Comment by Ajmdag** [comment]
I need this feature. How can I help with its implementation?

**Comment by rjmacarthy** [comment]
+1 for this please!

**Comment by silverqx** [comment]
Coders from vscode team can propose ideas on how this should be implemented, eg `Go to js implementation` or something like that, a first good step would be to create a ticket on vscode issue tracker and cross-reference this tickets so that communication can begin.

**Comment by weswigham** [comment]
> There's no easy way to navigate to the JS or TS implementation of a TS type if that type is declared in a DefinitelyTyped package. (or other cases where .d.ts isn't colocated with the implementation)

_Technically_ if you wanted to hand spin some sourcemaps to go with a DT package, it could kinda work out already. But there's definitely no automated solution here. Any solution we do have is based on improving scenarios involving TS compiler output, at those are the only times when we can otuput the additional metadata required to make the correct associations between files. Some packages support this today, I believe - I was using the azure js sdk the other day and debugging some stuff, and was pleasantly surprised when I went to some definitions and discovered they were in the original source. So yeah, it works with TS sources that have up-to-date `--declarationMaps` output bundled with them.

> Go To Type Definition doesn't work when the selection is itself a type. You'll get a "no type definition found" error.

I think y'all should open a new issue on that one. That's just a failure of exepections/bug, imo. Minimally, the error could be made better - something like "selection is only a type and so has no type definition associated with its value".

> Go To Definition on a non-type identifier behaves inconsistently. Sometimes it will navigate to the implementation, sometimes it won't. In theory this is the same as (1), but it's also possible to solve (1) while leaving this inconsistency in place.

It goes to the _first_ definition, where first is pretty arbitrary, and based on things like the order files were loaded in. The `peek` window is admittedly more useful, since it shows all definition sites. "Implementations" don't come into play.

**Comment by avkonst** [comment]
As a user, I will be satisfied, if it worked like this:

Ctrl+Click on a symbol (I mostly use only this to navigate the code) brings currently to either an implementation (if it is within the package being edited in VS Code) or to the declaration (if it is a package from the dependencies). The first does not require consideration. The second, when it goes to the declaration, I would like to Ctrl+Click on the declaration and it navigates me to the implementation, if it can locate, download and open it for me. Otherwise, it shows a message what problem prevented the sources to be found / opened. If it worked like this, it would be awesome.

I believe there are the following scenarios with locating the original source code:

1. A package is written in typescript and original source code in external repository and either
   a) distribution includes `d.ts` and `js` files OR
   b) distribution includes `d.ts`, `js` and `ts` files
2. A package is written in typescript and original source code in the current repository (mono repo case) and
   a) distribution includes `d.ts` and `js` files
3. A package is written in javascript and original source code in external repository and either
   a) distribution includes it's own `d.ts` and `js` files OR
   a) distribution includes only `js` files and external `@types` definitions are used
4. A package is written in javascript and original source code in the current repository (mono repo case) and
   a) distribution includes it's own `d.ts` and `js` files

1b) - is rarely used, and I do not like this approach as resources like bundlephobia and npm trends will report huge package sizes, which are not huge in runtime really, but people judge the size based on these resources. So, relying on the maintainers to include `ts` files with the distribution is not a good idea, but certainly if `ts` files are included, the case is solved.

1a) - (most?) frequent case. The package should some how declare (via package.json or source maps?)  where the original sources were located when compiled (like microsoft `pdb` symbol files declare) OR where can be downloaded from, for example git+revision+path. VS code would navigate to the local file or `git fetch` it to the local cache and open it on navigation from declaration to the implementation

2a) - if it is possible to detect the case of mono repo automatically (using some heuristics), the original Ctrl+Click navigation could open the implementation directly. If it is impossible, fallback to 1a) is good enough

I care less about 3, but it would be nice if 3a) could open JS implementation on the second Ctrl+Click navigation. In case of 3b) the `@types` package should declare what JS packages it annotates, so VS code could have this information to navigate to the JS implementation in the external package.

I care even less about 4, but it could have the same behaviour as 2a) - fallback for 4a) is 3a)

Hope it helps.

**Comment by weswigham** [comment]
>1a) - (most?) frequent case.

Yeah, that's what I'm saying we have all the tools in place to support now. We can combine `.js.map` and `.d.ts.map` files to produce a mapping from `.d.ts` to `.js` (without the intermediate `.ts`).  But, hm. Really, you'd prefer to go to the declaration file, and then interact again to go to the associated js? You'd not want to jump straight to the js?

**Comment by avkonst** [comment]
> "Really, you'd prefer to go to the declaration file, and then interact again to go to the associated js? You'd not want to jump straight to the js?"

In case of 1a), double jump or single jump are both fine for me. Although single jump can be considered as "more efficient", double jump has got the benefit of "reminding" the user that the second jump is a jump to the black box from the "library consumption" point of view (i.e. reminder that the source file being navigated to is not part of the current project). Maybe this behaviour could be made configurable.

In case of 3a), 3b) or 4, I **certainly** want double jump. I first want to see the declarations and only after this (in very rare cases, like debugging) the implementation. Jumping straight to JS skipping the TS declaration removes all the benefits of the declarations. So, I definitely want to see the TS declaration first.

**Comment by justingrant** [comment]
(combining replies to @weswigham and @avkonst from multiple posts)

> It goes to the first definition, where first is pretty arbitrary, and based on things like the order files were loaded in.

In the most common case where you want to go to the implementation for a particular imported symbol, why would there be multiple definitions?  Is this a common situation?

> produce a mapping from .d.ts to .js (without the intermediate .ts). 

Not sure I understand "without the intermediate TS". Do you mean that even if the original .ts source is available, VSCode would navigate to the transpiled .js?  That seems like a bad idea. If the original source is available, then VSCode should navigate there-- just like it would when the debugger steps into transpiled code that has a sourcemap. That said, if the package author was evil or lazy and didn't include the original TS source in their npm distro, then it seems reasonable for VSCode to fall back to the transpiled JS and perhaps show a warning message explaining why the user is seeing such weird, unreadable code and to encourage users to nag the package author to include the original source in their package.

> Really, you'd prefer to go to the declaration file, and then interact again to go to the associated js? You'd not want to jump straight to the js?

I don't think a two-step would add value. I think it's much better to have distinct, consistent UX: use "Go To Type Definition" to navigate to the type definition, and use Go To Definition (or Go To Implementation-- whatever we choose to call it) to go to the original source code or, per above, to the transpiled JS if original source isn't available. 

IMHO, it's much more confusing to have different behavior depending on whether the selected symbol is defined in your own code vs. in node_modules. It shouldn't matter. If I want to see the type, VSCode should take me to the type. If I want to see the implementation code, VSCode should take me there.  This is especially true because, in large organizations, what's "my code" and what's "library code" isn't always a bright line. If someone else refactored code into a shared internal package, I shouldn't have to change how I navigate to that code.

> I first want to see the declarations and only after this (in very rare cases, like debugging) the implementation.

Seems like it's clearer to require the developer to explicitly choose (by picking one menu item or another) whether they want to see a type declaration or to view the implementation code.  Having navigation be inconsistent depending on where the code comes from seems unnecessarily confusing.

> \> Go To Type Definition doesn't work when the selection is itself a type. 
> I think y'all should open a new issue on that one. That's just a failure of exepections/bug, imo. 

Yep, this could definitely be fixed first. I included it here because if my preferred solution ("Go to Definition always navigates to the implementation") is adopted, then users will get in the habit of using Go To Type Definition whenever they want to see the .d.ts, so it'd be weird if doing this on a type didn't act consistently to how it acts for a non-type identifier.

**Comment by avkonst** [comment]
> "It goes to the first definition, where first is pretty arbitrary, and based on things like the order files were loaded in."

I have not mentioned in my replies that it should navigate to the first definition. I said on Ctrl+Click on a symbol it should go *first of all* to the definition (as it works now) and only on the subsequent Ctrl+Click to the implementation.

**Comment by avkonst** [comment]
> "Seems like it's clearer to require the developer to explicitly choose (by picking one menu item or another) whether they want to see a type declaration or to view the implementation code."

Separate context menus are fine and good idea to have. They are not mutually exclusive to the Ctrl+Click behaviour.

**Comment by zjaml** [comment]
Wonder  why this critical bug is not being worked on yet. It is just becoming a more and more pain in the as as typescript adoption goes up.

**Comment by haugerbr** [comment]
@zjaml are you wanting it to go to TypeScript source files or JavaScript source files? TypeScript source files are supported. It's JavaScript source files that this issue is currently still open for.

**Comment by 0x80** [comment]
Tangentially related; I have a monorepo with a common package that's published to npm. Because of this public interface the dist is js + d.ts files. When I'm working in the repo VSCode always jumps to the d.ts instead of the source. Do you know a way around this?

**Comment by haugerbr** [comment]
@0x80 I believe this issue is still open for the scenario you just described. The only way I think you could get around this is if the project is open source and you compiled it from the typescript source. Then you could turn on the --declarationMaps flag for the typescript compiler and IDEs that support the maps will go to the typescript source instead of the d.ts files.

**Comment by cyberluke** [comment]
This is really crucial. I want simple thing: CTLR+Click and navigate to node_modules/library/source.js or source.ts file. Opensource project with 60 node_modules dependencies and you are back to using notepad with explorer.

**Comment by cvolant** [comment]
When I right-clicked on a component I saw: "Go to Definition" and "Go to Type Definition". I thought "Great, this has been well designed, the solution to my problem is just here!". Then I clicked on "Go to Definition" and... landed to the Type Definition.

**Comment by addnab** [comment]
The funky code navigation is the only reason why I'm still using Webstorm. I come back to Vscode once in a while to check if it's fixed. But it's not.

**Comment by snayam** [comment]
> This is really crucial. I want simple thing: CTLR+Click and navigate to node_modules/library/source.js or source.ts file. Opensource project with 60 node_modules dependencies and you are back to using notepad with explorer.

+1
Same need.

**Comment by timzekid** [comment]
Пофіксіть поскоріше, бо це треба наталі щоб робити верстку!!

**Comment by andrestone** [comment]
With all due respect, the fact that this isn't implemented is incredible. Such a basic, relatively easy to 
 build feature that would save people tons of time.

**Comment by or2008** [comment]
> With all due respect, the fact that this isn't implemented is incredible. Such a basic, relatively easy to
> build feature that would save people tons of time.

+1

**Comment by Happin3ss** [comment]
Well, in the meantime when this feature is not yet available, is there any better workaround beside searching for the whole node_modules ? It really inconvenience...
For example, I go to the definition and it led me to `node_modules/@com.abc/sample/lib/core/internal/abc.d.ts `when the one I want is actually located at `node_modules/@com.abc/sample/src/core/internal/abc.ts`

**Comment by justingrant** [comment]
The current solution AFAIK is [declaration maps](https://github.com/microsoft/TypeScript/pull/22658). These are sourcemaps that map .d.ts files to the corresponding TS source. If you can convince the libraries that you depend on to implement declaration maps (which I think effectively requires publishing .d.ts inside the library package as opposed to relying on @types) then you'll be able to automatically jump from your client code to the original TS (or even JS if using the TS compiler on JS?) source code.

What I've been doing is gradually filing PRs against my most important dependencies (at least the ones that are authored in TS) to get them to add declaration maps. 

Obviously it'd be better if changing the library weren't required.

**Comment by slikts** [comment]
Not sure if I've missed it, but it would be *really* useful for understanding what's happening to have actual minimal examples of sources, declarations and declaration maps. I've never seen either Go to Definition, Go to Type Definition, or Go to Implementations actually go to anything but a `.d.ts` file for transpiled TS. I just tried to make a [simple example](https://github.com/slikts/ts-go-to-def), and it doesn't work; I can't get VS Code to go to anything but the `.d.ts` file, even though the `.d.ts.map` is next to it. I've also found someone else [with the same issue](https://github.com/microsoft/TypeScript/issues/25662#issuecomment-457481516) who was ignored.

It's understandable that supporting external typings like `@types` would be hard, but what does it say that even the "happy path" can't be made to work?

There are *many* libraries that generate declaration files but not declaration maps, and it's unclear what the argument would be for enabling them. This option has been added without a clear explanation, at least one that I'd be able to find.

**Comment by avkonst** [comment]
Yes, I confirm that declaration maps are not solving the problem. My project has got declaration maps but vscode does not navigate beyond .d.ts files.

**Comment by zachweinberg** [comment]
It sure would be great if this happened.

**Comment by russeg** [comment]
if there is an "elephant in the room" of issues for ms vscode devs, this is it.

**Comment by belykh** [comment]
Any updates on this one?
I just decided to make a big step: move from JS to TS. I thought - "what am I doing wrong?" but actually, it's unexpected behavior of VSCode

**Comment by Happin3ss** [comment]
> Any updates on this one?
> I just decided to make a big step: move from JS to TS. I thought - "what am I doing wrong?" but actually, it's unexpected behavior of VSCode

It was a long history since this behavior had been raised. Well, nothing going on here. No plan so far.
 For me, I keep using vscode as a kind of editor as it is, and Intellij products (IDEA) for my coding stuffs even I don't like it so much.

**Comment by DorianMaliszewski** [comment]
+1 For this feature

**Comment by YegorZaremba** [comment]
Opened for 5 years, hope it will be fixed in 2025

**Comment by sb-bilal-dev** [comment]
Just another miserable comment from ts dev going through hell everyday to find definitions

**Comment by parksj10** [comment]
This is crazy... I thought I was just a TS noob, but now I see that the behavior I’m getting is actually “standard”???? So frustrating, every other IDE/language I’ve used has this implemented at initial release... :( I love VS code but honesty this could be a deal breaker

**Comment by Zx-EvM** [comment]
Can anyone explain why `go to definition/implementation` features in VSCode with Python extensions for, huh Python language works better then for JS having native built-in solutions?

**Comment by prbxr** [comment]
So what is going on?
I am using scss.d.ts for css modules and can't jump to scss.

Also when creating simple .js and accordingly d.ts, I cannot jump straight to js

**Comment by xiancanhui** [comment]
why not use webstorm?
it can solve this problem!!!

**Comment by Yegorich555** [comment]
> why not use webstorm?
> it can solve this problem!!!
> my version 2020.1

Because it's not open-source solution and not free

**Comment by ftlno** [comment]
Please fix this.

**Comment by silverqx** [comment]
It's absolutely clear, that this problem can be fixed only if some manager from Microsoft takes over, this problem has to be managed and coordinated, a manager develops a workflow and assigns work to coders. ☝
This problem has to be sent to the management to take over and they decide how it will be implemented.
So if is here somebody competent from Microsoft, he can give a shot. 😊

**Comment by pikeas** [comment]
+1, the absence of this feature makes it very difficult to be productive in a Typescript codebase. VS Code provides "Go to definition", "go to type definition", and "go to implementations". Currently, all three do what I expect from "go to type definition", and none of them go to the implementation.

**Comment by cnbebop** [comment]
Can it be fixed in 2021 ?

**Comment by Happin3ss** [comment]
> Can it be fixed in 2021 ?
 Based on my 4+ years experience following this issue, NO
 My suggestion for you again, is change your IDE to Jetbrains's if that feature is essential for you (like what I did). Otherwise, just accept it as it's the down side for a free and opens source product 😄 Your voice is listened, but not always.

**Comment by ThaJay** [comment]
I'm joining the party. Maybe in this case noise and jokes on GitHub are okay.

Any workarounds apart from manually navigating or project wide search?

**Comment by ThaJay** [comment]
Today, I got the latest VsCode update and it has become a bigger problem now. Now, When I do Go To Definition on a React Component, it goes to `node_modules\@types\react\index.d.ts:547` for EVERY component.

I'm switching editors for typescript if this won't be fixed soon. This feature is highly needed, I use it hundreds of times per day.

**Comment by gino8080** [comment]
> Today, I got the latest VsCode update and it has become a bigger problem. Now, When I do Go To Definition on a React Component, it goes to `node_modules\@types\react\index.d.ts:547` for EVERY component.
> 
> I'm switching editors for typescript if this won't be fixed soon. This feature is highly needed, I use it hundreds of times per day.
> 
> Edit: It seems to be a tradeoff, now it does go to definition instead of type definition with objects and if you go to definition on the import of a component, it does show the implementation instead of the type. So it's still workable but it also still needs more improvement.

yes is VERY ANNOYING,
i want to go to component declaration and not the type definitions
vscode older version was working

**Comment by andrewmartin** [comment]
@ThaJay I'm seeing this one as well, was kind of annoying and strange. The thing that's most disappointing is that it worked great for many months.

I wonder if there is any way to tweak things in settings to get this to work?

**Comment by andrewmartin** [comment]
One workaround I found is to do `CMD + F12` instead, which will go right to the implementation.

**Comment by wtfzambo** [comment]
> 
> 
> Today, I got the latest VsCode update and it has become a bigger problem. Now, When I do Go To Definition on a React Component, it goes to `node_modules\@types\react\index.d.ts:547` for EVERY component.
> 
> I'm switching editors for typescript if this won't be fixed soon. This feature is highly needed, I use it hundreds of times per day.
> 
> Edit: It seems to be a tradeoff, now it does go to definition instead of type definition with objects and if you go to definition on the import of a component, it does show the implementation instead of the type. So it's still workable but it also still needs more improvement.

Did you manage to find a way to go to proper definition? I'm using react and whatever piece of imported code I try to get the definition of, it no matter what sends me to `node_modules\@types\react\index.d.ts`.

**Comment by ThaJay** [comment]
@wtfzambo I use goto definition on a variable that is not in JSX and it works most of the time (like the import statement or somewhere else in the file). But clicking a component name in a  JSX part leads me to `node_modules\@types\react\index.d.ts`

**Comment by wtfzambo** [comment]
@ThaJay ah crap, nvm then, I have everything in JSX so I guess I'm out of luck for the time being >.<

**Comment by andrewmartin** [comment]
@ThaJay @wtfzambo Just click on the component once in the editor, then press `CMD + F12`. Best workaround for the time being.

**Comment by wtfzambo** [comment]
@andrewmartin I tried that, but it still sends me to `index.d.ts`. See:

![image](https://user-images.githubusercontent.com/47462792/115527366-48c8aa00-a291-11eb-94f2-51fe5e8f025b.png)` ==> `![image](https://user-images.githubusercontent.com/47462792/115529404-2b94db00-a293-11eb-87cd-10e93ca37ced.png)

But then at this point I'm thinking I am not fully understanding how this whole thing works, since coming from python, if I F12 upon anything I'd just get sent to the source code of whatever I'm using.

**Comment by Zx-EvM** [comment]
The one of must have features of each IDE is 6 year in backlog still?

**Comment by ThaJay** [comment]
@andrewmartin Thank you, I understand now.

Nice that it has been fixed. On to the actual question then, I want to set my ctrl+click gesture to goto implementation and ctrl+f12 shortcut to goto definition.

There does not seem to be an option to change the command for the ctrl+click gesture.

**Comment by andrewmartin** [comment]
@wtfzambo it looks to me like you're trying to go to the definition of a `<div>` element. `<div>` comes from React. Try putting your cursor on the word `<Navbar />` in the _consumer_ (wherever you use `Navbar` in your app), then press `CMD + F12`.

@Zx-EvM this worked fine before and still technically works. Just need a way to configure the `CTRL + CLICK` gesture as @ThaJay just mentioned. I think it may be possible actually, so I'm digging in a little more...

**Comment by wtfzambo** [comment]
@andrewmartin you're right sorry, I should have been more specific. It works for custom components (such as my `<Navbar />`) but it doesn't work for stuff that I import from `node_modules`.

But I guess I was expecting a python-like behavior that in JS is not really a thing, as one of my colleagues explained to me.

**Comment by justingrant** [comment]
@wtfzambo - my understanding is that when sourcemaps are involved (as they will usually be for imports from `node_modules`) then declaration maps are also needed in addition to sourcemaps. A "declaration map" enables connecting a TS declaration to the implementation, even if the implementation is in JS which is sourcemapped back to TS.

See https://stackoverflow.com/questions/58459988/what-is-the-use-case-of-enabling-declaration-and-declarationmap-in-tsconfig-json for details.

Relatively few npm packages have declaration maps enabled. ;-(

**Comment by gaurav5430** [comment]
@andrewmartin cmd+f12 (Go to implementations) does not work for me, it still takes me to the type definition file or just says "No implementations available" . Is there another setting that needs to be configured to make this work?

**Comment by christophe-g** [comment]
So, this is the reason why this works when a file is named **`source.js`** (it brings me to the definition)

![image](https://user-images.githubusercontent.com/1168053/126341181-2b551ee3-72dc-4038-adc4-ee5334959652.png)

And does not when it is called **`source.ts`**?

Really, after 7 years!

**Comment by gaurav5430** [comment]
> So, this is the reason why this works when a file is named **`source.js`** (it brings me to the definition)
> 
> ![image](https://user-images.githubusercontent.com/1168053/126341181-2b551ee3-72dc-4038-adc4-ee5334959652.png)
> 
> And does not when it is called **`source.ts`**?
> 
> Really, after 7 years!
> 
> 
> 

what is the reason exactly?

**Comment by ThaJay** [comment]
> But I guess I was expecting a python-like behavior that in JS is not really a thing, as one of my colleagues explained to me.

It should be a thing. If it's not a thing (it usually isn't) then that's because of things not working like intended. I had a js project once where it all worked and it was glorious.

**Comment by mankey-ru** [comment]
Can confirm this is very annoying. "Go to definition" in VSCode works same as "Go to type definition"

**Comment by Robula** [comment]
I'm not sure if this case is about jumping to the compiled JS or the original TS file, the former I see as being less useful. Usually I want to jump to the TS source file instead of the generated d.ts.

**Comment by justingrant** [comment]
> I'm not sure if this case is about jumping to the compiled JS or the original TS file, the former I see as being less useful. Usually I want to jump to the TS source file instead of the generated d.ts.

The ideal behavior would be for the IDE to navigate to the "original source", meaning the file pointed to by the sourcemap of the transpiled JS. Sometimes this will be a TS file, sometimes a JS file that's been transpiled, or could even be some other language that's been transpiled down to JS. 

I agree that showing the transpiled JS is not as helpful, and could be actively confusing for some transpilations (e.g. of async/await) which are more-or-less incomprehensible for newbies... and sometimes for experts too!

**Comment by sabetAI** [comment]
Is there anything blocking a community extension to get around this issue?

**Comment by ThaJay** [comment]
> I'm not sure if this case is about jumping to the compiled JS or the original TS file, the former I see as being less useful. Usually I want to jump to the TS source file instead of the generated d.ts.

I just want to go to exactly the place where the import points. Nothing else is even remotely interesting compared to this dead simple link. If the one file says `import {Foo} from ../Stuff` and I follow `Foo`, I want to go one folder up, into the file Stuff and highlighting the line that says `export Foo`.

It seems quite simple and I rather not have typescript at all than this not working.

**Comment by mifi** [comment]
GitHub UI supports click to jump to implementation OR type definition. I wonder what trickery they are using...

<img width="435" alt="Screenshot 2021-09-21 at 12 50 52" src="https://user-images.githubusercontent.com/402547/134118734-3dd264fc-1542-4ed1-b191-ac3342af21ea.png">

**Comment by sb-bilal-dev** [comment]
I might soon ditch VSCode in favor of Github UI then😄

**Comment by publicvirtualvoid** [comment]
Is this ticket not in the right place?
Is everyone else happy manually trawling through node_modules every time they want to look at an implementation or is there something I'm missing?

**Comment by phil294** [comment]
> Is this ticket not in the right place? Is everyone else happy manually trawling through node_modules every time they want to look at an implementation or is there something I'm missing?

No, you're not missing anything, everybody is indeed very unhappy with the current situation (there's 100 hidden items above)

**Comment by silverqx** [comment]
Every project has tickets like this @publicvirtualvoid that is absolutely normal, I can name 10 tickets like this right away right now in other projects 😂, maybe not 10 right now, but you know what I mean 🙂

**Comment by sabetAI** [comment]
@silverqx at least maintainers in other projects have better outreach (ie Flutter, even with all its issues). If they could explain their rationale and all the technical roadblocks, I'm sure the community could help in **some** way. It's been six years already for such a fundamental feature though, so a solution is probably never coming. It's a pretty embarrassing situation for their team to be in tbh.

**Comment by kevinwang930** [comment]
@sabetAI  I think the reasonable explanation is that go to implementation has nothing to do with typescript, and as a typescript team they definitely want everyone use typescript, so they has less willingness to improve the javascript user experience.

**Comment by justingrant** [comment]
I'm happy to see continued enthusiasm for solving this problem. Keep the upvotes and (especially) constructive suggestions and PRs coming! Like many hard things, eventually I'm sure we'll help to get a solution built for this annoying gap in the TS developer experience.

> @sabetAI I think the reasonable explanation is that go to implementation has nothing to do with typescript, and as a typescript team they definitely want everyone use typescript, so they has less willingness to improve the javascript user experience. Actually if you delete all the type declaration files go to implementation works.

As far as I can tell, this impression is not correct. Almost all* packages that are written in TS, transpiled to JS, and published to npm don't work with "Go To Implementation" any more than packages originally written in JS.

I said "almost all" above because there is a way to make it work: using the [declarationMap](https://www.typescriptlang.org/tsconfig) tsconfig compiler option.  But:

1. This setting is not enabled by default.  IMHO it should be enabled by default whenever `sourceMap` is enabled. (Maybe also when `inlineSourceMap` is enabled?)
2. There is AFAICT no evangelism from the TS team (or anyone else!) about how package authors should enable this setting and include declaration map files when publishing their packages to npm.
3. There's also no push to get declaration map files into existing DefinitelyTyped packages. Could this be automated instead of relying on 1000s of type-package maintainers to do it?

Anyone who wants to see this problem get better could:

1. **Send a PR to try to change the default of `declarationMap` to be `true` if the `sourceMap` or `inlineSourceMap` option is `true`**. This minor change would probably have the biggest impact on this problem over time.

2. **Send a PR to update the TS docs to be more vehement that `declarationMap` should be enabled for all library packages**, and explaining why ("_because "Go To Implementation" in VS Code relies on declaration map files_"). These documentation changes will be crucial if (1) is not approved.

3. **Send a PR that adds or extends the TS docs with specific guidance for package authors**, including "publish your sourcemaps" and "publish declaration maps". 

4. **Write a tool that will automatically add declaration map files to all compatible DefinitelyTyped packages**, and convince the DefinitelyTyped maintainers to merge the output of that tool into every compatible DefinitelyTyped package.  Part of the work would be figuring out what "compatible" means and what packages' declaration maps can't be automated.  Note that @weswigham above was pessimistic that this automation would be possible. But I wonder if things have improved since then to make automation possible? Or if it'd be possible for a large subset of packages? Here's what he said:

> There's no easy way to navigate to the JS or TS implementation of a TS type if that type is declared in a DefinitelyTyped package. (or other cases where .d.ts isn't colocated with the implementation)
> 
> Technically if you wanted to hand spin some sourcemaps to go with a DT package, it could kinda work out already. But there's definitely no automated solution here. Any solution we do have is based on improving scenarios involving TS compiler output, at those are the only times when we can otuput the additional metadata required to make the correct associations between files. Some packages support this today, I believe - I was using the azure js sdk the other day and debugging some stuff, and was pleasantly surprised when I went to some definitions and discovered they were in the original source. So yeah, it works with TS sources that have up-to-date --declarationMaps output bundled with them.
 

5. **Send a PR to change the build actions for DefinitelyTyped to warn users if they try to add a new package that lacks a declaration map**.  Not sure if this is something that can be done outside the DT maintainer team, depending on how the existing DT toolchain works. Someone who knows DT better than me could answer this.

The community has agency!  If we want things to get better, any of the items above can make things better. 

There's also IMHO another confusion with "Go to Definition" vs. "Go To Type Definition" are implemented. My preference would be for "Go to Type Definition" to navigate to the types and "Go to Definition" to navigate to the implementation. Not sure this is worth changing community expectations, though, given how long the current behavior has been in place. For more details, see: https://github.com/microsoft/TypeScript/issues/6209#issuecomment-525911553

**Comment by justingrant** [comment]
> > Go To Type Definition doesn't work when the selection is itself a type. You'll get a "no type definition found" error.
> 
> I think y'all should open a new issue on that one. That's just a failure of exepections/bug, imo. Minimally, the error could be made better - something like "selection is only a type and so has no type definition associated with its value".

@weswigham - sorry for delayed reply, I just opened #46627 for this.

**Comment by andrewbranch** [comment]
Thanks @justingrant! I want to make a few observations.

If I understand correctly, the thing in highest demand here is to be able to jump (whether as an explicit action or automatically under the hood as part of an existing UI action) from a location in a `.d.ts` file to a location in a `.js` file (and perhaps a `.ts` file is also an acceptable target for some people if it exists). The declaration maps and source maps generated by the TypeScript compiler allow you to jump from a location in an output file (`.js` or `.d.ts`) to a location in an input file (typically `.ts`, but could be `.js`).

On top of this, I’ll add some assertions that I think most people would agree with:

- Most npm libraries written in TypeScript do not publish their `.ts` sources to npm; they publish only the `.js` and `.d.ts` outputs.
- DefinitelyTyped packages often do not track versions 1:1 with their implementation libraries; DT is updated asynchronously by volunteers, and furthermore, many patch version bumps of the implementation libraries do not require updates to DT as these patches typically don't change the public API surface.

Putting all this together, I can draw a few intermediate conclusions:

- Even if it were possible to auto-generate declaration maps for DefinitelyTyped, it would not be practical because the chances are extremely low that any user who obtains that declaration map would have the exact version of the implementation library that it was generated against. Every time the implementation library changes anything, even just adding a newline somewhere, the DT package would have to be regenerated against that source, and users would have to update them in lockstep or else they’d get sent to an inaccurate file / line number. This is a total non-starter.
- Turning our attention away from the DefinitelyTyped problem and toward libraries written in TypeScript: enabling and shipping declaration maps alone would not change anything; those libraries would need to ship the declaration maps _and_ the `.ts` source files in order for us to use the declaration maps. If libraries started doing that, we could easily jump from the `.d.ts` to the `.ts` (or skip the `.d.ts` altogether—maybe this is already what happens?). Even in that case, though, we would have no way of getting to the `.js` source. I think _some_ people would be happy just to see the `.ts` source, but I can also imagine pushback at the idea of making the world download TypeScript source files from npm just to light up an editor feature. I think _more_ people would be just as happy or happier to be able to go to the `.js` and leave the `.ts` out of their node_modules. But nothing we’ve discussed so far is remotely useful in making this happen.

Because of these problems, I’m actually a little confused at the focus of this issue on declaration maps. It seems like they would help only in edge case scenarios, and I think I can categorically reject the idea that we could use them to map from DefinitelyTyped to another source. The only solution I can think of that could solve both the problem both for DefinitelyTyped and for bundled type definitions is for us to make an effort to find, parse, and bind the corresponding JS file and make an effort to find the symbol in question. This would surely fail in the face of highly dynamic or unusual JS patterns, or where `.d.ts` files don't cleanly map 1:1 to `.js` files. It could also be expensive in the face of huge bundled and minified files. This would have to be understood as a heuristic-based best effort, and perhaps a worst-case fallback would be to take you to the top of the package’s `"main"` (or relevant `"exports"`) file. But I also suspect that this process would work perfectly for a lot of libraries, and personally, I think just opening up the `"main"` entrypoint would save me time over scrolling through hundreds of node_modules folders when I really need to track something down in the source.

**Comment by justingrant** [comment]
Hi @andrewbranch - thanks for jumping in on this important issue. After following it for a few years, I'll add context that may help. 

* My understanding is that Go To Implementation currently works, but only for packages that publish their .d.ts, sourcemaps, declaration maps, and original sources to npm. Unless I'm mistaken, packages that are compiled with TS (an increasing number of popular packages!) can make Go To Implementation work today.
  * Caveat: TSC may emit declaration maps that work, but popular transpilers and bundlers (rollup, esbuild, babel, etc.) may not. `rollup-plugin-ts` [has problems](https://github.com/wessberg/rollup-plugin-ts/issues/117), and I suspect others do too. If the TS team could help those maintainers fix their support, the ecosystem would be hugely grateful!
* Increasingly, maintainers *are* packaging original sources, because original sources (or inline sourcemaps) are required for easier debugging in browser dev tools and IDEs like VSCode. 
  * This is especially true for the increasing number of popular packages that are being ported to TS.
  * I've pushed PRs to fix sourcemaps and ship original source for [10+ packages](https://github.com/pulls?q=is%3Apr+author%3Ajustingrant+is%3Apublic+%28%22source+map%22+OR+%22sourcemap%22%29+). If my memory is correct, only once has a maintainer refused to ship source. The others were happy to ship source because it helped their users debug their package more easily. 
  * There are ecosystem changes coming (e.g. to [create-react-app](https://github.com/facebook/create-react-app/pull/8227), perhaps later to [`npm publish`](https://github.com/justingrant/rfcs/blob/latest/accepted/0000-validate-sourcemaps.md)) that may accelerate the trend of shipping original source along with sourcemaps.
* Transpiled JS is often hard to understand, esp. for async/await or other ES6+ features that don't transpile cleanly to ES5. For this reason, the VSCode debugger's Step Into (the debugger counterpart of Go To Implementation) always only shows transpiled JS if the original source is unavailable. Seems like Go To Implementation should behave the same way, both for consistency across different ways of "navigating" and also because many users are confused by transpiled source.
* You're 💯  correct that DT won't work with declaration maps because of version mismatches. My fault for not thinking this through before posting.
  * I'm seeing many package maintainers bundling their .d.ts with their packages to avoid versioning mismatches, or because they're porting their source to TS. There's still a *lot* of reliance on DT, so this trend doesn't invalidate the need for a DT-friendly solution. But it also suggests that the non-DT (bundled declarations) case is important too and will get more important over time. 

For all the reasons above, I suspect that the right solution is two-tracked:

1. For packages built with TS, packaging declaration maps (along with sourcemaps, .d.ts, and original sources) seems like a good solution, because it can provide granular, precise navigation that I'm not sure is possible with an exports-only solution. I assume it'd provide the best perf too. Also it actually may work today!  But...
   * We're [lacking guidance](https://github.com/microsoft/TypeScript/issues/16792#issuecomment-934499824) for package maintainers about how to make declaration maps, sourcemaps, etc. work correctly. IMO, what's needed is general ""how to build a library in TS" docs. Here's a possible [outline](https://github.com/microsoft/TypeScript/issues/16792#issuecomment-892068076) for this content.
   * As noted above, non-TSC bundlers and transpilers probably need help with emitting declaration maps properly because today those maps can be broken by code splitting, tree shaking, etc.
   * Consider defaulting `declarationMap` to be `true` if the `sourceMap` or `inlineSourceMap` option is `true`
   * For packages with inline sourcemaps, can we use the same code that the JS debugger extension uses to show "sources" using inline sourcemaps?
 
2. For packages that aren't built with TS (or where the declaration map => sourcemap => original source chain is broken), there should be a fallback solution.  For example assume the following code:
```js
// mode_modules/some-package/src/Foo.js
import doSomething from 'do-something';
class Foo {
  static bar(x) {
    doSomething(x);
  }
}
export default Foo;

// mayapp.ts
import Foo from 'some-package';
const x = Foo.bar(42);
```

I assume that Go To Implementation for `import('some-package').Foo.bar`  should behave as follows:
* If the location of `bar` can be resolved, e.g. via sourcemaps or some TS JIT magic, then navigate to `static bar(x) {`
* If all we know is imports, then navigate to `class Foo {` (I'm assuming that the same magic that the IDE uses to make Go To Definition work for regular JS files will also work as long as  `Foo` is declared at global scope.)
* If all we know is the file (e.g. if package.json has `exports`) but the exported symbol cant be resolved in that file, then navigate to the first line of Foo.js
* If we don't even know the file then navigate to `main` or `module`.

The bullet-points above are pretty hand-wavey because I don't know enough about TS and Node's module-resolution behavior. But I assume TS and/or Node-resolution experts will know what's possible!

@andrewbranch what do you think?  You likely understand this stuff better than I do, so please correct me where I'm wrong above.

**Why navigating to transpiled source is bad**

> * I think _more_ people would be just as happy or happier to be able to go to the `.js` and leave the `.ts` out of their node_modules.

Are you sure about this?  Have you heard this sentiment from median-skill developers?  I'm skeptical, because transpiled code can be really hard to understand for non-advanced developers.

For example, here's an example of the code I saw in VSCode when I first debugged into a [method](https://unpkg.com/browse/@aws-amplify/auth@4.3.12/lib-esm/Auth.js#L439) of [AWS Amplify](https://github.com/aws-amplify/amplify-js), Amazon's JS front-end library that wraps AWS services:

```js
           onSuccess: function (session) { return __awaiter(_this, void 0, void 0, function () {
                var cred, e_1, currentUser, e_2;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            logger.debug(session);
                            delete user['challengeName'];
                            delete user['challengeParam'];
                            _a.label = 1;
                        case 1:
                            _a.trys.push([1, 4, 5, 9]);
                            return [4 /*yield*/, this.Credentials.clear()];
                        case 2:
                            _a.sent();
                            return [4 /*yield*/, this.Credentials.set(session, 'session')];
                        case 3:
                            cred = _a.sent();
                            logger.debug('succeed to get cognito credentials', cred);
                            return [3 /*break*/, 9];
                        case 4:
                            e_1 = _a.sent();
                            logger.debug('cannot get cognito credentials', e_1);
                            return [3 /*break*/, 9];
                        case 5:
                            _a.trys.push([5, 7, , 8]);
                            return [4 /*yield*/, this.currentUserPoolUser()];
                        case 6:
                            currentUser = _a.sent();
                            that.user = currentUser;
                            dispatchAuthEvent('signIn', currentUser, "A user " + user.getUsername() + " has been signed in");
                            resolve(currentUser);
                            return [3 /*break*/, 8];
                        case 7:
                            e_2 = _a.sent();
                            logger.error('Failed to get the signed in user', e_2);
                            reject(e_2);
                            return [3 /*break*/, 8];
                        case 8: return [7 /*endfinally*/];
                        case 9: return [2 /*return*/];
                    }
                });
            }); },
```

Eek!  I assume that most JS developers won't easily understand happening in the state machine above. But after https://github.com/aws-amplify/amplify-js/pull/2680 was merged, debugging into the same method showed the [original source](https://github.com/aws-amplify/amplify-js/blob/main/packages/auth/src/Auth.ts#L533), which is much easier to understand:

```js
			onSuccess: async session => {
				logger.debug(session);
				delete user['challengeName'];
				delete user['challengeParam'];
				try {
					await this.Credentials.clear();
					const cred = await this.Credentials.set(session, 'session');
					logger.debug('succeed to get cognito credentials', cred);
				} catch (e) {
					logger.debug('cannot get cognito credentials', e);
				} finally {
					try {
						// In order to get user attributes and MFA methods
						// We need to trigger currentUserPoolUser again
						const currentUser = await this.currentUserPoolUser();
						that.user = currentUser;
						dispatchAuthEvent(
							'signIn',
							currentUser,
							`A user ${user.getUsername()} has been signed in`
						);
						resolve(currentUser);
					} catch (e) {
						logger.error('Failed to get the signed in user', e);
						reject(e);
					}
				}
			},
```

I agree that there are some cases where sophisticated users may want to see the transpiled JS. But those are unusual cases. The normal case (and the only case that many unsophisticated developers can handle) is to show the original source, for the same reasons that we show original source in the debugger.

> But I also suspect that this process would work perfectly for a lot of libraries, and personally, I think just opening up the `"main"` entrypoint would save me time over scrolling through hundreds of node_modules folders when I really need to track something down in the source.

I think you are correct here. It doesn't have to be perfect to add a lot of value.

**Comment by andrewbranch** [comment]
Thanks for the detailed explanation, and all good points. I don’t doubt that publishing TS sources to npm for debugging purposes is becoming more common. Likewise, as you also noted, we’re seeing a [slow but steady trend](https://twitter.com/sanders_n/status/1419691599936720896) of packages moving away from DefinitelyTyped to ship their own types (presumably most of these are autogenerated from TS sources, but some are hand-authored or generated from a tool other than tsc). However, I feel confident that the majority of people’s node_modules _today_ lacks TS implementation files, so any solution that does not include an attempt to resolve symbols to JS, at least when no other option is present, I think will be highly unsatisfying and confusing.

> Are you sure about this? Have you heard this sentiment from median-skill developers? I'm skeptical, because transpiled code can be really hard to understand for non-advanced developers.

Nope! This was a total guess based on past experience with JS-only VS Code users who get confused (and sometimes angry 😅) when they get dropped into something that looks TypeScripty. I think there _are_ users who would rather get dropped into the JS, but at this point I don’t have evidence to say they make up a particular ratio of all users. But, I don’t think it really matters—_if_ we have all the pieces in place to jump to the exact right place in a `.ts` file, we can be nearly 100% sure that result is correct and we can calculate it very fast, so I think we should always do that. The point I should have been making was, again, that we need to be able to make a guess at the JS when the TS doesn’t exist.

I guess my thought process was, if I need to look at implementation files in node_modules, something has already gone horribly, horribly wrong, and I’m bracing for a bad time. I hopefully don’t need to fully understand generator downleveling in order to find a line to put a breakpoint on. But I’m sure my standards here are too low for long-term thinking 😄. It seems like you’re doing a very good job of thinking about how to improve things for the long game, which I very much appreciate. I was admittedly more focused on how to get acceptable results now. Point well taken that we should be thinking about both. It’s probably worth at least doing some writing about the benefits of publishing declaration maps and sources with your packages.

> The bullet-points above are pretty hand-wavey because I don't know enough about TS and Node's module-resolution behavior. But I assume TS and/or Node-resolution experts will know what's possible!

I actually don’t think we know what’s possible without some experiments. My suspicion is that exports will usually be easy, and drilling down into them further will be hit-or-miss, but I don’t know.

**Comment by justingrant** [comment]
Cool, sounds good. A few responses inline. So glad you're looking into this!

> Nope! This was a total guess based on past experience with JS-only VS Code users who get confused (and sometimes angry 😅) when they get dropped into something that looks TypeScripty.

Yep. Those users have a difficult tradeoff: hard-to-read transpiled code or also-hard-to-read (for them) TS code. But this is exactly the same problem they have in the debugger today, and debugging into node_modules is *much* more common than using Go To Implementation for node_modules.

For this reason, I'd recommend thinking of these problems separately. 
1. How can Go To Implementation show original source (from source/declaration maps) with a fallback to transpiled JS if the mapping doesn't work. The default will be "original source" because that's consistent with how (more common) debugger use cases behave.
2. How to easily flip from transpiled source to original source (and flip back too) in the IDE?  This would affect the most common case (the debugger) as well as less common cases like Go To Implementation. FWIW, when debugging I frequently have cases where I want to view transpiled source (e.g. when I suspect the sourcemap is out-of-date after a hot reload) so this feature would be useful beyond TS haters.  The UI could be a right-click command, or even a button in the upper-right-corner toolbar like the button used to navigate from a diff view to the underlying file.

I think keeping the problems separate will make it easier to make progress on both in parallel without getting stuck in arguments about whether to show TS or JS for Go To Implementation. 

Question: do declaration maps that TSC emits point to transpiled code or to original source?  If the former, then they'll work just like the debugger. 👍  But if the latter, then it seems like it'd be hard to find the transpiled JS. I guess TS could load the entire sourcemap of the package and see if the file/row/col that was resolved by the declaration map also shows up in a sourcemap file, and trace it back from there?  Anyway, it'd be good to know whether declaration maps point to transpiled or original source.

> _if_ we have all the pieces in place to jump to the exact right place in a `.ts` file, we can be nearly 100% sure that result is correct and we can calculate it very fast, so I think we should always do that. The point I should have been making was, again, that we need to be able to make a guess at the JS when the TS doesn’t exist.

👍  Yep!  I come back to aligning with the (more commonly used for node_modules navigation) debugger behavior: if there's original source, show that. If not, then try to fall back to something reasonable.

> if I need to look at implementation files in node_modules, something has already gone horribly, horribly wrong, and I’m bracing for a bad time.

Many years ago, I was working with the VS team back when the original "Just My Code" feature was being built. There was a big debate between the C# and (especially) the C++ team whose users had a long history of viewing, debugging, and copying libraries like MFC vs. the VB team whose users were confused by library code, esp. because libraries were written in C# and C++!  What's nice is that both preferences have been improved. For library-code-haters, Just My Code continues to exist and be enhanced. For library-code-lovers, it's gotten easier to view, debug into, and modify library code.

But the fact that 15 years later this debate hasn't gone away is a signal that this is a persistent difference between developers. For some devs, viewing library code is an awful, exceptional case. But for many other users, debugging, viewing, and even editing/PR-ing library code is a normal course of business, because libraries are often buggy, poorly documented, or need to be extended in some way.  You see this parallel in other walks of life too. For example, if a light switch isn't working, some folks will grab some tools while others will grab a phone and call an electrician.

So I think it's helpful to assume that for many developers (but not all!) they will read, debug into, fix, and borrow library code frequently, so we should try to make it work as well as possible.

For Just My Code fans, they won't be clicking Go To Implementation anyways (because their preference is to avoid library code!) so I think this feature is really targeted only at the anti-JMC crowd.

> exports will usually be easy, and drilling down into them further will be hit-or-miss, but I don’t know.

Yep, I think the same. 

> It’s probably worth at least doing some writing about the benefits of publishing declaration maps and sources with your packages.

What are next steps to get more attention on this?  In https://github.com/microsoft/TypeScript/issues/16792#issuecomment-892068076 I wrote a possible outline for this content. I'm happy to help contribute but honestly I feel unqualified to write it because I don't understand enough about the problem space. But I'd love to pair with someone to help make this content happen. LMK how I could help.

A good place to start might be to simply creating a "Writing libraries in TypeScript" section in the docs and putting some minimal content there. Then that section in the docs could be a honeypot to attract tips and tricks from various TS library authors via PRs. In addition to sourcemaps/decl-maps content, there could be sections about how to configure various bundlers, how to migrate from DT to shipping types, and other topics that only apply to library authors.

**Comment by danieliser** [comment]
So one thing I haven't seen mentioned, though I may have missed it in this long thread, I miss seeing the actual code comments, function & parameter descriptions. Sure types are great, but knowing the package authors intention for each function is almost more valuable.

`date-fns` for example doesn't tell you in their `.d.ts` files which date is supposed to come after the other for functions like `isAfter`. The answer is clear in the functions actual Docblocks. And digging into a large node_modules folder to find that one js file is tedious when you need to do it a dozen times.

What VSCode gives me:

![image](https://user-images.githubusercontent.com/847818/150441933-fb1b1eae-253f-4ea6-916e-b089912e94d7.png)

What the package authors intended me to see:
![image](https://user-images.githubusercontent.com/847818/150441889-9c6eb733-e300-41bb-bc2e-e341eaf25e53.png)

**Comment by andrewbranch** [comment]
@danieliser totally agree—however, if we get a solution to this at all, it’s likely to be a “best guess” effort to find the corresponding JS from the typings. I’m ok with providing a go-to function for that because it’s better than nothing, but I don’t think we’ll achieve high enough confidence (or satisfactory performance) in that guess in order to pull in the JSDoc. This seems like a case where the answer is just to get the package authors to copy the JSDoc comments to the typings (or copy it in a DT PR). If the typings are generated from TS source, the comments will be copied to the typings automatically. If they’re missing, it means someone’s manual process or third-party tooling is insufficient and the problem should be addressed there.

**Comment by andrewbranch** [comment]
Please note the milestone change means I plan to investigate solutions to this during the 4.7 timeframe. I don’t think we have a clear path yet, so it doesn’t necessarily mean that something will ship in 4.7.

**Comment by mhofman** [comment]
I'd like to add a use case: some of our files are authored as `.js` files (with JSDoc style typing) with explicit/manual `.d.ts` files for them to express things that cannot be expressed otherwise in JSDoc (e.g. circular type definitions, function overloads, interface, etc.)

The problem is that with such files, it's impossible to "go to definition" / "go to implementation" as VS Code and other editors stop at the `.d.ts` file. `declarationMap` isn't an option since the `.d.ts` file is manually authored.

**Comment by andrewbranch** [comment]
I have a prototype up at #48264 that I’m welcoming feedback on. I’m 100% sure it doesn’t cover everything but if folks are willing to give it a try, it would be helpful to gather some test cases that don’t work yet so I can see if there’s a good way forward on them.

To try it: There’s a build from @typescript-bot on the PR (copied below) that you can npm install, then ensure your VS Code window is using that build of TypeScript by opening the command palette (in a TS or JS file) and running `TypeScript: Select TypeScript Version`. You should see the one newly in your node_modules listed as version 4.7.0. At that point, your Go To Definition command will be replaced with the new logic that finds more JS files and ignores .d.ts files. 🚨 **This is not the final experience! We are not replacing Go To Definition like this!** We don’t know for sure yet where the new logic will be exposed: it might be in a new command, or its results might be merged into Go To Definition or Go To Implementation. There are technical and UX reasons why the latter might not be desirable, but if it’s in a new command, none of you would be able to test it right now without also running a custom build of VS Code. So this build **temporarily** replaces the existing go-to-def implementation with the new command so you can easily test it.

One specific thing to keep an eye out for in testing: There are some cases where we are quite confident in the results we return, other cases where the best we can do is jumping to the top of a file, and then other cases where we’re confident about the file but we make a number of guesses within the file. Under the hood, I am indicating whether or not the result is a guess or not, but this information isn’t currently exposed in VS Code. If we go forward with presenting guesses like this, it will almost certainly go along with some kind of visual indicator noting that it’s a guess. Here’s an example on `import { add } from "lodash"`:

<img width="823" alt="image" src="https://user-images.githubusercontent.com/3277153/158282106-64ee6d75-a71f-49e7-9ae0-66b5a5141c65.png">

Two of these guesses are right, while the top one is wrong. If you notice more places containing wrong guesses like this, please note that and tell me how to repro them on the PR. I need a bit of crowdsourcing to determine whether the guesses are worth doing at all and if there’s an easy way to improve them.

Thanks all!

> Hey @andrewbranch, I've packed this into [an installable tgz](https://typescript.visualstudio.com/cf7ac146-d525-443c-b23c-0d58337efebc/_apis/build/builds/121842/artifacts?artifactName=tgz&fileId=BC8F78CBCF3D3B551A34EA6659FA5D670D458543ADE83DD287314B98C2CEFF7A02&fileName=/typescript-4.7.0-insiders.20220315.tgz). You can install it for testing by referencing it in your `package.json` like so:
> ```
> {
>     "devDependencies": {
>         "typescript": "https://typescript.visualstudio.com/cf7ac146-d525-443c-b23c-0d58337efebc/_apis/build/builds/121842/artifacts?artifactName=tgz&fileId=BC8F78CBCF3D3B551A34EA6659FA5D670D458543ADE83DD287314B98C2CEFF7A02&fileName=/typescript-4.7.0-insiders.20220315.tgz"
>     }
> }
> ```
> and then running `npm install`.

**Comment by andrewbranch** [comment]
The plan is to add an experimental new command in VS Code to trigger the behavior added in #48264. I’ll keep this thread updated to let you know when you can try it out in VS Code Insiders + typescript@next.

**Comment by justingrant** [comment]
Great progress here. Thanks @andrewbranch for your perseverance in making this happen!  Can't wait to try it out.

**Comment by andrewbranch** [comment]
> I’ll keep this thread updated to let you know when you can try it out in VS Code Insiders + typescript@next.

I forgot 🤦‍♂️

You can try this in VS Code Insiders and typescript@next! And pretty soon it will be in the stable versions. I started an issue to gather feedback: #49003. Thanks!

**Comment by selrond** [comment]
Already in the newest release: [Visual Studio Code April 2022](https://code.visualstudio.com/updates/v1_67#_typescript-47-support)

**Comment by rshea0** [comment]
@andrewbranch Thank you thank you **thank you** for making this happen! You just made my life so much easier
